### ISSUE TITLE
includeSoftConstraints configuration param not being respected for TopologySpreadConstraint

### ISSUE THREAD

**Issue by vholer** [kubernetes-sigs/descheduler#960]
**What version of descheduler are you using?**

descheduler version: 0.25.0

**Does this issue reproduce with the latest release?**

yes

**Which descheduler CLI options are you using?**

`--policy-config-file /policy-dir/policy.yaml --descheduling-interval 30s --v 5`

**Please provide a copy of your descheduler policy config file**

```
apiVersion: "descheduler/v1alpha1"
kind: "DeschedulerPolicy"
strategies:
  LowNodeUtilization:
    enabled: false
  PodLifeTime:
    enabled: false
  RemoveDuplicates:
    enabled: false
  RemovePodsHavingTooManyRestarts:
    enabled: false
  RemovePodsViolatingInterPodAntiAffinity:
    enabled: false
  RemovePodsViolatingNodeAffinity:
    enabled: false
  RemovePodsViolatingNodeTaints:
    enabled: false
  RemovePodsViolatingTopologySpreadConstraint:
    enabled: true
    params:
      includeSoftConstraints: true
      nodeFit: true
```

**What k8s version are you using (`kubectl version`)?**

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.4", GitCommit:"95ee5ab382d64cfe6c28967f36b53970b8374491", GitTreeState:"archive", BuildDate:"2022-08-18T00:00:00Z", GoVersion:"go1.18.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.3", GitCommit:"aef86a93758dc3cb2c658dd9657ab4ad4afc21cb", GitTreeState:"clean", BuildDate:"2022-07-13T14:23:26Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}
</pre></details>

**What did you do?**

I have deployed 6 pods across 2 nodes with topologySpreadConstraints:

```
      topologySpreadConstraints:
        - maxSkew: 1
          topologyKey: kubernetes.io/hostname
          whenUnsatisfiable: ScheduleAnyway
          labelSelector:
            matchLabels:
              role: app
              size: small
```

The pods are deployed/balanced nicely:

```
$ kubectl get nodes
$ kubectl get nodes
NAME           STATUS                     ROLES           AGE    VERSION
minikube       Ready                      control-plane   3h9m   v1.24.3
minikube-m02   Ready,SchedulingDisabled   <none>          3h8m   v1.24.3
minikube-m03   Ready                      <none>          3h8m   v1.24.3

$ kubectl get pods --no-headers -o wide | awk '{ print $7, $1 }' | sort
minikube app-small-2-66b7679d9-lzz5p
minikube app-small-4-79744df667-w5qf6
minikube app-small-6-7f74cd4c4f-bbn8f
minikube-m03 app-small-1-64d5db69f8-fw2hx
minikube-m03 app-small-3-5dc48dd4d4-vxcqf
minikube-m03 app-small-5-78b7f54c9d-qd6vz
```

I have uncordoned third node and expected the pods to rebalance:

```
$ kubectl get nodes
NAME           STATUS   ROLES           AGE    VERSION
minikube       Ready    control-plane   3h9m   v1.24.3
minikube-m02   Ready    <none>          3h8m   v1.24.3
minikube-m03   Ready    <none>          3h7m   v1.24.3
```

but descheduler didn't notice them since they are configured with `whenUnsatisfiable: ScheduleAnyway` and descheduler doesn't respect `RemovePodsViolatingTopologySpreadConstraint.params.includeSoftConstraints`:

```
I0926 11:30:56.034429       1 healthz.go:174] Installing health checkers for (/healthz): "Descheduler"
I0926 11:30:56.035331       1 named_certificates.go:53] "Loaded SNI cert" index=0 certName="self-signed loopback" certDetail="\"apiserver-loopback-client@1664191856\" [serving] validServingFor=[apiserver-loopback-client] issuer=\"apiserver-loopback-client-ca@1664191855\" (2022-09-26 10:30:55 +0000 UTC to 2023-09-26 10:30:55 +0000 UTC (now=2022-09-26 11:30:56.035271017 +0000 UTC))"
I0926 11:30:56.035383       1 secure_serving.go:210] Serving securely on [::]:10258
I0926 11:30:56.035467       1 tlsconfig.go:240] "Starting DynamicServingCertificateController"
I0926 11:30:56.059491       1 reflector.go:221] Starting reflector *v1.Pod (0s) from k8s.io/client-go/informers/factory.go:134
I0926 11:30:56.059556       1 reflector.go:257] Listing and watching *v1.Pod from k8s.io/client-go/informers/factory.go:134
I0926 11:30:56.059557       1 reflector.go:221] Starting reflector *v1.Namespace (0s) from k8s.io/client-go/informers/factory.go:134
I0926 11:30:56.059595       1 reflector.go:257] Listing and watching *v1.Namespace from k8s.io/client-go/informers/factory.go:134
I0926 11:30:56.060499       1 reflector.go:221] Starting reflector *v1.PriorityClass (0s) from k8s.io/client-go/informers/factory.go:134
I0926 11:30:56.060535       1 reflector.go:257] Listing and watching *v1.PriorityClass from k8s.io/client-go/informers/factory.go:134
I0926 11:30:56.159518       1 shared_informer.go:285] caches populated
I0926 11:30:56.159778       1 shared_informer.go:285] caches populated
I0926 11:30:56.159887       1 shared_informer.go:285] caches populated
I0926 11:30:56.163328       1 node.go:49] "Node lister returned empty list, now fetch directly"
I0926 11:30:56.180707       1 descheduler.go:343] Building a pod evictor
I0926 11:30:56.186221       1 topologyspreadconstraint.go:115] "Processing namespaces for topology spread constraints"
I0926 11:30:56.226491       1 descheduler.go:426] "Number of evicted pods" totalEvicted=0
I0926 11:31:05.052735       1 pathrecorder.go:241] descheduler: "/healthz" satisfied by exact match
I0926 11:31:15.051492       1 pathrecorder.go:241] descheduler: "/healthz" satisfied by exact match
I0926 11:31:25.051514       1 pathrecorder.go:241] descheduler: "/healthz" satisfied by exact match
I0926 11:31:26.164185       1 node.go:49] "Node lister returned empty list, now fetch directly"
I0926 11:31:26.180062       1 descheduler.go:343] Building a pod evictor
I0926 11:31:26.185711       1 topologyspreadconstraint.go:115] "Processing namespaces for topology spread constraints"
I0926 11:31:26.220488       1 descheduler.go:426] "Number of evicted pods" totalEvicted=0
I0926 11:31:35.051024       1 pathrecorder.go:241] descheduler: "/healthz" satisfied by exact match
I0926 11:31:45.051353       1 pathrecorder.go:241] descheduler: "/healthz" satisfied by exact match
```

**What did you expect to see?**

I have expected the 2 of 6 pods to be evicted and scheduled on newly uncordoned third node.

**What did you see instead?**

Nothing nappened.

**Comment by vholer** [comment]
Bug is on place initializing the plugins with wrong argument:

https://github.com/kubernetes-sigs/descheduler/blob/e31f4aac009b3d79fd76736f5fb3f33008e56a11/pkg/descheduler/strategy_migration.go#L215

instead of with `params.IncludeSoftConstraints`.

With the fix of above, I get pod constraints evaluated correctly and 2 expected pods being evicted:

```
I0926 13:37:07.225469  111537 topologyspreadconstraint.go:115] "Processing namespaces for topology spread constraints"
I0926 13:37:07.235530  111537 node.go:147] "Pod fits on node" pod="default/app-small-6-7f74cd4c4f-bbn8f" node="minikube-m02"
I0926 13:37:07.235596  111537 node.go:147] "Pod fits on node" pod="default/app-small-5-78b7f54c9d-qd6vz" node="minikube-m02"
I0926 13:37:07.258599  111537 node.go:49] "Node lister returned empty list, now fetch directly"
I0926 13:37:07.265194  111537 node.go:147] "Pod fits on node" pod="default/app-small-5-78b7f54c9d-qd6vz" node="minikube"
I0926 13:37:07.275959  111537 evictions.go:160] "Evicted pod" pod="default/app-small-5-78b7f54c9d-qd6vz" reason="" strategy="RemovePodsViolatingTopologySpreadConstraint" node="minikube-m03"
I0926 13:37:07.276103  111537 node.go:49] "Node lister returned empty list, now fetch directly"
I0926 13:37:07.299778  111537 node.go:147] "Pod fits on node" pod="default/app-small-6-7f74cd4c4f-bbn8f" node="minikube-m02"
I0926 13:37:07.323780  111537 evictions.go:160] "Evicted pod" pod="default/app-small-6-7f74cd4c4f-bbn8f" reason="" strategy="RemovePodsViolatingTopologySpreadConstraint" node="minikube"
I0926 13:37:07.323859  111537 descheduler.go:431] "Number of evicted pods" totalEvicted=2
I0926 13:37:37.214677  111537 node.go:49] "Node lister returned empty list, now fetch directly"
I0926 13:37:37.224199  111537 descheduler.go:348] Building a pod evictor
I0926 13:37:37.228282  111537 topologyspreadconstraint.go:115] "Processing namespaces for topology spread constraints"
I0926 13:37:37.237668  111537 topologyspreadconstraint.go:203] "Skipping topology constraint because it is already balanced" constraint={MaxSkew:1 TopologyKey:kubernetes.io/hostname WhenUnsatisfiable:ScheduleAnyway LabelSelector:&LabelSelector{MatchLabels:map[string]string{role: app,size: small,},MatchExpressions:[]LabelSelectorRequirement{},} MinDomains:<nil> NodeAffinityPolicy:<nil> NodeTaintsPolicy:<nil> MatchLabelKeys:[]}
I0926 13:37:37.263884  111537 descheduler.go:431] "Number of evicted pods" totalEvicted=0
```

and

```
$ kubectl get pods --no-headers -o wide | awk '{ print $7, $1 }' | sort
minikube app-small-2-66b7679d9-lzz5p
minikube app-small-4-79744df667-w5qf6
minikube-m02 app-small-5-78b7f54c9d-n9w9l
minikube-m02 app-small-6-7f74cd4c4f-q4hw9
minikube-m03 app-small-1-64d5db69f8-fw2hx
minikube-m03 app-small-3-5dc48dd4d4-vxcqf
```

**Comment by damemi** [comment]
Hi @vholer, thanks for reporting this. It should be fixed in the `v0.25.1` patch release, which we just published. If that works, then we'll close this issue.

**Comment by vholer** [comment]
Hello @damemi, I do confirm that with `v0.25.1` release the descheduler correctly evicts the `ScheduleAnyway` pods when `includeSoftConstraints` is enabled. You can close the issue, thanks!
